datafusion/functions-nested/src/max.rs

```diff
@@ -0,0 +1,173 @@
+        Self::new()
+    }
+}
+
+impl ArrayMax {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::array(Volatility::Immutable),
+            aliases: vec!["list_max".to_string()],
+        }
+    }
+}
+
+impl ScalarUDFImpl for ArrayMax {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "array_max"
+    }
+
+    fn display_name(&self, args: &[Expr]) -> datafusion_common::Result<String> {
+        let args_name = args.iter().map(ToString::to_string).collect::<Vec<_>>();
+        if args_name.len() != 1 {
+            return exec_err!("expects 1 arg, got {}", args_name.len());
+        }
+
+        Ok(format!("{}", args_name[0]))
+    }
+
+    fn schema_name(&self, args: &[Expr]) -> datafusion_common::Result<String> {
+        let args_name = args
+            .iter()
+            .map(|e| e.schema_name().to_string())
+            .collect::<Vec<_>>();
+        if args_name.len() != 1 {
+            return exec_err!("expects 1 arg, got {}", args_name.len());
+        }
+
+        Ok(format!("{}", args_name[0]))
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> datafusion_common::Result<DataType> {
+        match &arg_types[0] {
+            List(field) | LargeList(field) | FixedSizeList(field, _) => {
+                Ok(field.data_type().clone())
+            }
+            _ => exec_err!(
+                "Not reachable, data_type should be List, LargeList or FixedSizeList"
+            ),
+        }
+    }
+
+    fn invoke_batch(
+        &self,
+        args: &[ColumnarValue],
+        _number_rows: usize,
+    ) -> datafusion_common::Result<ColumnarValue> {
+        make_scalar_function(array_max_inner)(args)
+    }
+
+    fn aliases(&self) -> &[String] {
+        &self.aliases
+    }
+
+    fn documentation(&self) -> Option<&Documentation> {
+        self.doc()
+    }
+}
+
+/// array_max SQL function
+///
+/// There is one argument for array_max as the array.
+/// `array_max(array)`
+///
+/// For example:
+/// > array_max(\[1, 3, 2]) -> 3
+pub fn array_max_inner(args: &[ArrayRef]) -> datafusion_common::Result<ArrayRef> {
+    if args.len() != 1 {
+        return exec_err!("array_max needs one argument");
+    }
+
+    match &args[0].data_type() {
+        List(_) | LargeList(_) | FixedSizeList(_, _) => {
+            let new_args = vec![
+                args[0].clone(),
+                Arc::new(StringArray::from_iter(vec![Some("DESC")])),
+                Arc::new(StringArray::from_iter(vec![Some("NULLS LAST")])),
+            ];
+            array_max_internal(&new_args)
+        }
+        _ => exec_err!("array_max does not support type: {:?}", args[0].data_type()),
+    }
+}
+
+fn array_max_internal(args: &[ArrayRef]) -> datafusion_common::Result<ArrayRef> {
+    let sorted_array = array_sort_inner(args)?;
+    let result_array = as_list_array(&sorted_array)?.value(0);
```

Comment:
Sorting is more expensive than just finding the max.

Reply:
did you try removing the sorting step?

Reply:
Can we reuse `max()` agg's code here?
https://github.com/apache/datafusion/blob/48a28af801cc9144fbd8e8c076deb360c9e5f49b/datafusion/functions-aggregate/src/min_max.rs#L577

Reply:
Yes, it is applicable by supporting all requirements of array_max. Addressed.